### ISSUE TITLE
keras_mnist.ipynb

### ISSUE THREAD

**Issue by OoallenoO123123123** [OoallenoO123123123/AI-#2]
# 匯入必要的模組
from tensorflow import keras
from keras.models import Sequential
from keras.layers import Dense, Flatten
import matplotlib.pyplot as plt
----------------------------------------------------------------------------------------------------------------
# 載入 MNIST 手寫數字資料集
# MNIST 共有 60000 筆訓練資料、10000 筆測試資料
mnist=keras.datasets.mnist
(train_data, train_label), (test_data, test_label) = mnist.load_data()

Downloading data from https://storage.googleapis.com/tensorflow/tf-keras-datasets/mnist.npz
11490434/11490434 [==============================] - 2s 0us/step
-----------------------------------------------------------------------------------------------------------------
# 顯示資料的型態與維度
print(type(train_data))  # numpy 陣列
print(train_data.shape)  # (60000, 28, 28)：60000 張 28x28 的灰階圖
print(type(train_label))
print(train_label.shape) # (60000,)：每張圖的標籤 (0~9)
print(test_data.shape)  # (10000, 28, 28)
print(test_label.shape) # (10000,)

<class 'numpy.ndarray'>
(60000, 28, 28)
<class 'numpy.ndarray'>
(60000,)
(10000, 28, 28)
(10000,)
----------------------------------------------------------------------------------------------------------------
# 顯示第一筆資料的標籤與影像矩陣
print(train_label[0])  # 顯示該影像代表的數字
print(train_data[0])  # 顯示該影像的像素值 (0~255)

5
[[  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   3  18  18  18 126 136
  175  26 166 255 247 127   0   0   0   0]
 [  0   0   0   0   0   0   0   0  30  36  94 154 170 253 253 253 253 253
  225 172 253 242 195  64   0   0   0   0]
 [  0   0   0   0   0   0   0  49 238 253 253 253 253 253 253 253 253 251
   93  82  82  56  39   0   0   0   0   0]
 [  0   0   0   0   0   0   0  18 219 253 253 253 253 253 198 182 247 241
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0  80 156 107 253 253 205  11   0  43 154
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0  14   1 154 253  90   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0 139 253 190   2   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0  11 190 253  70   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0  35 241 225 160 108   1
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0  81 240 253 253 119
   25   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0  45 186 253 253
  150  27   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0  16  93 252
  253 187   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0 249
  253 249  64   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0  46 130 183 253
  253 207   2   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0  39 148 229 253 253 253
  250 182   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0  24 114 221 253 253 253 253 201
   78   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0  23  66 213 253 253 253 253 198  81   2
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0  18 171 219 253 253 253 253 195  80   9   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0  55 172 226 253 253 253 253 244 133  11   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0 136 253 253 253 212 135 132  16   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]]
---------------------------------------------------------------------------------------------------------
import matplotlib.pyplot as plt
plt.imshow(train_data[0], cmap='binary')
plt.show()

# 建立神經網路模型
model=Sequential()
model.add(Flatten())  # 將 28x28 的影像展平成一維（784 個輸入特徵）

# 第一層：全連接層（32 個神經元，ReLU 激活函數）
# input_dim=784 表示輸入向量長度
model.add(Dense(32, activation='relu', input_dim=784))  

# 第二層（輸出層）：10 個神經元（對應 0~9），sigmoid 激活函數
model.add(Dense(10, activation='sigmoid'))

# 編譯模型
# optimizer: 使用 RMSprop 最佳化器
# loss: 使用 sparse_categorical_crossentropy（處理整數標籤）
# metrics: 追蹤準確率
model.compile(optimizer='rmsprop', loss='sparse_categorical_crossentropy', metrics=['accuracy'])

# 訓練模型
# epochs=20：訓練 20 回合
# batch_size=512：每批處理 512 張影像
model.fit(train_data,train_label,epochs=20,batch_size=512)

# 使用測試資料評估模型表現
score=model.evaluate(test_data, test_label)

# 輸出測試集準確率
print(score[1])  # 約 0.93 (即 93.2%)

Epoch 1/20
118/118 [==============================] - 8s 6ms/step - loss: 6.6168 - accuracy: 0.3339
Epoch 2/20
118/118 [==============================] - 1s 6ms/step - loss: 1.0995 - accuracy: 0.7135
Epoch 3/20
118/118 [==============================] - 1s 7ms/step - loss: 0.7904 - accuracy: 0.8069
Epoch 4/20
118/118 [==============================] - 1s 7ms/step - loss: 0.6363 - accuracy: 0.8471
Epoch 5/20
118/118 [==============================] - 1s 6ms/step - loss: 0.5263 - accuracy: 0.8725
Epoch 6/20
118/118 [==============================] - 1s 6ms/step - loss: 0.4557 - accuracy: 0.8856
Epoch 7/20
118/118 [==============================] - 1s 12ms/step - loss: 0.3942 - accuracy: 0.8975
Epoch 8/20
118/118 [==============================] - 1s 10ms/step - loss: 0.3528 - accuracy: 0.9053
Epoch 9/20
118/118 [==============================] - 1s 8ms/step - loss: 0.3148 - accuracy: 0.9122
Epoch 10/20
118/118 [==============================] - 1s 6ms/step - loss: 0.2978 - accuracy: 0.9171
Epoch 11/20
118/118 [==============================] - 1s 6ms/step - loss: 0.2767 - accuracy: 0.9228
Epoch 12/20
118/118 [==============================] - 1s 5ms/step - loss: 0.2597 - accuracy: 0.9267
Epoch 13/20
...
Epoch 20/20
118/118 [==============================] - 0s 3ms/step - loss: 0.1945 - accuracy: 0.9435
313/313 [==============================] - 1s 3ms/step - loss: 0.2882 - accuracy: 0.9322
0.932200014591217
